cmd/prow-job-dispatcher/main.go

```diff
@@ -61,9 +83,248 @@ func (o *options) complete(secrets *sets.String) error {
+			cloudProviders.Insert(ct)
+		}
+	}
+	return cloudProviders
+}
+
+type clusterVolume struct {
+	// [cloudProvider][cluster]volume
+	clusterVolumeMap map[string]map[string]float64
+	// only needed for stable tests: traverse the above map by sorted key list
+	cloudProviders sets.String
+	mutex          sync.Mutex
+}
+
+// findClusterForJobConfig finds a cluster running on a preferred cloud provider for the jobs in a Prow job config.
+// The chosen cluster will be the one with minimal workload with the given cloud provider.
+// If the cluster provider is empty string, it will choose the one with minimal workload across all cloud providers.
+func (cv *clusterVolume) findClusterForJobConfig(cloudProvider string, jc *prowconfig.JobConfig, path string, config *dispatcher.Config, jobVolumes map[string]float64) string {
+	//no cluster in the build farm is from the targeting cloud provider
+	if _, ok := cv.clusterVolumeMap[cloudProvider]; !ok {
+		cloudProvider = ""
+	}
+	var cluster, rCloudProvider string
+	min := float64(-1)
+	cv.mutex.Lock()
+	for _, cp := range cv.cloudProviders.List() {
+		m := cv.clusterVolumeMap[cp]
+		for c, v := range m {
+			if cloudProvider == "" || cloudProvider == cp {
+				if min < 0 || min > v {
+					min = v
+					cluster = c
+					rCloudProvider = cp
+				}
+			}
+		}
+	}
+
+	for k := range jc.PresubmitsStatic {
+		for _, job := range jc.PresubmitsStatic[k] {
+			cv.addToVolume(rCloudProvider, cluster, job.JobBase, path, config, jobVolumes)
+		}
+	}
+
+	for k := range jc.PostsubmitsStatic {
+		for _, job := range jc.PostsubmitsStatic[k] {
+			cv.addToVolume(rCloudProvider, cluster, job.JobBase, path, config, jobVolumes)
+		}
+	}
+	for _, job := range jc.Periodics {
+		cv.addToVolume(rCloudProvider, cluster, job.JobBase, path, config, jobVolumes)
+	}
+
+	cv.mutex.Unlock()
+	return cluster
+}
+
+func (cv *clusterVolume) addToVolume(cloudProvider, cluster string, jobBase prowconfig.JobBase, path string, config *dispatcher.Config, jobVolumes map[string]float64) {
+	determinedCluster, canBeRelocated := config.DetermineClusterForJob(jobBase, path)
+	if cluster == string(determinedCluster) || canBeRelocated {
+		cv.clusterVolumeMap[cloudProvider][cluster] = cv.clusterVolumeMap[cloudProvider][cluster] + jobVolumes[jobBase.Name]
+	} else if determinedCloudProvider, ok := config.IsInBuildFarm(determinedCluster); ok {
+		cv.clusterVolumeMap[string(determinedCloudProvider)][string(determinedCluster)] = cv.clusterVolumeMap[string(determinedCloudProvider)][string(determinedCluster)] + jobVolumes[jobBase.Name]
+	}
+}
+
+// dispatchJobConfig dispatches the jobs defined in a Prow jon config
+func (cv *clusterVolume) dispatchJobConfig(jc *prowconfig.JobConfig, path string, config *dispatcher.Config, jobVolumes map[string]float64) string {
+	cloudProvidersForE2ETests := getCloudProvidersForE2ETests(jc)
+	var cluster string
+	if cloudProvidersForE2ETests.Len() == 1 {
+		cloudProvider, _ := cloudProvidersForE2ETests.PopAny()
```

Comment:
Ah, this got me confused because of the `PopAny` which will obviously return the one and only result